go/obscuronode/enclave/enclave.go

```diff
@@ -326,71 +346,88 @@ func (e *enclaveImpl) produceRollup(b *types.Block, bs *obscurocore.BlockState)
-		speculativeRollup := <-e.speculativeWorkOutCh
-
-		newRollupTxs = speculativeRollup.txs
-		newRollupState = speculativeRollup.s
-		newRollupHeader = speculativeRollup.h
-
-		// the speculative execution has been processing on top of the wrong parent - due to failure in gossip or publishing to L1
-		// or speculative execution is disabled
-		speculativeExecutionSucceeded = speculativeRollup.found && (speculativeRollup.r.Hash() == bs.HeadRollup)
-
-		if !speculativeExecutionSucceeded && speculativeRollup.r != nil {
-			nodecommon.LogWithID(e.nodeShortID, "Recalculate. speculative=r_%d(%d), published=r_%d(%d)",
-				obscurocommon.ShortHash(speculativeRollup.r.Hash()),
-				speculativeRollup.r.Header.Number,
-				obscurocommon.ShortHash(bs.HeadRollup),
-				headRollup.Header.Number)
-			if e.statsCollector != nil {
-				e.statsCollector.L2Recalc(e.nodeID)
+	/*
+			speculativeExecutionSucceeded := false
+		   todo - reenable
+			if e.speculativeExecutionEnabled {
+				// retrieve the speculatively calculated State based on the previous winner and the incoming transactions
+				e.speculativeWorkInCh <- true
+				speculativeRollup := <-e.speculativeWorkOutCh
+
+				newRollupTxs = speculativeRollup.txs
+				newRollupState = speculativeRollup.s
+				newRollupHeader = speculativeRollup.h
+
+				// the speculative execution has been processing on top of the wrong parent - due to failure in gossip or publishing to L1
+				// or speculative execution is disabled
+				speculativeExecutionSucceeded = speculativeRollup.found && (speculativeRollup.r.Hash() == bs.HeadRollup)
+
+				if !speculativeExecutionSucceeded && speculativeRollup.r != nil {
+					nodecommon.LogWithID(e.nodeShortID, "Recalculate. speculative=r_%d(%d), published=r_%d(%d)",
+						obscurocommon.ShortHash(speculativeRollup.r.Hash()),
+						speculativeRollup.r.Header.Number,
+						obscurocommon.ShortHash(bs.HeadRollup),
+						headRollup.Header.Number)
+					if e.statsCollector != nil {
+						e.statsCollector.L2Recalc(e.nodeID)
+					}
+				}
 			}
+	*/
+
+	successfulTransactions := make([]nodecommon.L2Tx, 0)
+	// if !speculativeExecutionSucceeded {
+	// In case the speculative execution thread has not succeeded in producing a valid rollup
+	// we have to create a new one from the mempool transactions
+	newRollupHeader = obscurocore.NewHeader(&bs.HeadRollup, headRollup.Header.Number+1, e.nodeID)
+	newRollupTxs = currentTxs(headRollup, e.mempool.FetchMempoolTxs(), e.storage)
+	newRollupState = e.storage.CreateStateDB(bs.HeadRollup)
+	receipts := evm.ExecuteTransactions(newRollupTxs, newRollupState, newRollupHeader, e.storage, e.chainID)
+	// todo - only transactions that fail because of the nonce should be excluded
+	for _, tx := range newRollupTxs {
+		_, f := receipts[tx.Hash()]
+		if f {
+			successfulTransactions = append(successfulTransactions, tx)
+		} else {
+			log.Info(">   Agg%d: Excluding transaction %d", obscurocommon.ShortAddress(e.nodeID), obscurocommon.ShortHash(tx.Hash()))
```

Comment:
@tudor-malene V minor, but should we indicate that the tx was excluded for failing to execute correctly? Is there any info we can provide on why?

Reply:
weird, I thought I was printing the error. I'll double check.

Reply:
actually, the error is printed when it is executed.